src/variable_info.hpp

```diff
@@ -16,98 +16,119 @@
 
 #pragma once
 
-/** Information on a WML variable. */
-#include <string>
 #include "config.hpp"
 #include "variable_info_detail.hpp"
 
-class invalid_variablename_exception : public std::exception
+#include <string>
+#include <type_traits>
+
+// Restrict alias to local scope.
+namespace
 {
-public:
-	invalid_variablename_exception() : std::exception() {}
-	const char* what() const NOEXCEPT
-	{
-		return "invalid_variablename_exception";
-	}
-};
+namespace impl = variable_info_implementation;
+}
 
-template<const variable_info_detail::variable_info_type vit>
+/** Information on a WML variable. */
+template<typename vit>
 class variable_info
 {
 public:
+	variable_info(const std::string& varname, typename impl::maybe_const<config, vit>::reference vars) NOEXCEPT;
 
-	typedef typename variable_info_detail::maybe_const<vit,config>::type config_var;
-	/// Doesn't throw
-	variable_info(const std::string& varname, config_var& vars);
-	~variable_info();
 	std::string get_error_message() const;
-	/// Doesn't throw
-	bool explicit_index() const;
-	/// might throw invalid_variablename_exception
+
+	bool explicit_index() const NOEXCEPT;
+
+	/** @throws                 invalid_variablename_exception */
 	bool exists_as_attribute() const;
-	/// might throw invalid_variablename_exception
+
+	/** @throws                 invalid_variablename_exception */
 	bool exists_as_container() const;
 
 	/**
-		might throw invalid_variablename_exception
-		NOTE:
-			If vit == vit_const, then the lifime of the returned const attribute_value& might end with the lifetime of this object.
-	*/
-	typename variable_info_detail::maybe_const<vit, config::attribute_value>::type &as_scalar() const;
-	/// might throw invalid_variablename_exception
-	typename variable_info_detail::maybe_const<vit, config>::type & as_container() const;
-	/// might throw invalid_variablename_exception
-	typename variable_info_detail::maybe_const<vit, config::child_itors>::type as_array() const; //range may be empty
+	 * If vit == vi_policy_const, then the lifetime of the returned const attribute_value& might end
+	 * with the lifetime of this object.
+	 *
+	 * @throws                  invalid_variablename_exception
+	 */
+	typename impl::maybe_const<config::attribute_value, vit>::reference as_scalar() const;
+
+	/** @throws                 invalid_variablename_exception */
+	typename impl::maybe_const<config, vit>::reference as_container() const;
```

Comment:
"Jyrki thought they should take an rvalue reference."

I don't follow what exactly you're talking about. This comment chain is about a function that doesn't take *any* parameters.

I have mentioned earlier that the functions in `variable_info_mutable` should receive the vector *either* by value *or* by rvalue reference, but I don't see how that's relevant to @CelticMinstrel's comment at all.